src/babel-serializer/geometry/obj.cc

```diff
@@ -1,2 +1,417 @@
+        prev_pos = pos;
+        if (pos >= data.size())
+        {
+            on_error(data, data.last(0), "expected line", severity::error);
+        }
+
+        // skip comment lines
+        while (pos < data.size() && char(data[pos]) == '#')
+        {
+            while (pos < data.size() && char(data[pos]) != '\n')
+                ++pos;
+            if (pos < data.size() && char(data[pos]) == '\n')
+                ++pos;
+        }
+
+        auto p_start = reinterpret_cast<char const*>(data.data() + pos);
+        while (pos < data.size())
+        {
+            auto c = char(data[pos]);
+            if (c == '\n' || c == '#')
+                break;
+
+            ++pos;
+        }
+        auto p_end = reinterpret_cast<char const*>(data.data() + pos);
+
+        // ignore rest of comment
+        while (pos < data.size())
+        {
+            auto c = char(data[pos]);
+            if (c == '\n')
+                break;
+
+            ++pos;
+        }
+
+        // point behind newline
+        if (pos < data.size() && char(data[pos]) == '\n')
+            ++pos;
+
+        // trim whitespaces
+        while (pos < data.size() && cc::is_space(char(data[pos])))
+            ++pos;
+
+        return cc::string_view(p_start, p_end);
+    };
+
+    auto const has_more_lines = [&]() -> bool { return pos < data.size(); };
+
+    auto const parse_int = [&](cc::string_view s) -> int {
+        cc::string ss = s; // null-terminated
+        int i;
+        auto const r = std::sscanf(ss.c_str(), "%d", &i);
+        if (r == 0)
+        {
+            on_error(data, cc::as_byte_span(s), "unable to parse int", severity::error);
+            return 0;
+        }
+        return i;
+    };
+
+    auto const parse_float = [&](cc::string_view s) -> ScalarT {
+        cc::string ss = s; // null-terminated
+        ScalarT f;
+        if constexpr (std::is_same_v<ScalarT, float>)
+        {
+            auto const r = std::sscanf(ss.c_str(), "%f", &f);
+            if (r == 0)
+            {
+                on_error(data, cc::as_byte_span(s), "unable to parse float", severity::error);
+                return 0;
+            }
+        }
+        else
+        {
+            auto const r = std::sscanf(ss.c_str(), "%lf", &f);
+            if (r == 0)
+            {
+                on_error(data, cc::as_byte_span(s), "unable to parse double", severity::error);
+                return 0;
+            }
+        }
+        return f;
+    };
+
+    auto const parse_vertex = [&](cc::string_view line) {
+        CC_ASSERT(line[0] == 'v');
+        CC_ASSERT(cc::is_space(line[1]));
+        tg::pos<4, ScalarT> p;
+        p.w = ScalarT(1);
+        int i = 0;
+        for (auto const s : line.subview(2).split(cc::is_space, cc::split_options::skip_empty))
+        {
+            if (i > 3)
+            {
+                on_error(data, cc::as_byte_span(line), "unable to parse vertex", severity::error);
```

Comment:
`on_error` might not throw / abort (maybe just log an error / warning and handle it later). Thus you need to break here (otherwise `p[i++]` will be out of bounds). Also affects other parse_xyz functions.

Reply:
Ah good point! I was also unsure what things should call `on_error` and which should be added to the unrecognized lines.

Reply:
unrecognized should be unrecognized "line type", e.g. `vc ` if someone wants to extend obj with such vertex colors
errors in normal types (e.g. `vn`) should call on_error

Reply:
Alright, gotcha! :+1: